feat: model status refresh interval and conditions

havasik opened the pull request:
Related to #37
Fixes #35

Features:
- Periodic model validation and health checking in the controller
- Model status tracking with Kubernetes conditions (Ready, Discovering, Error)
- Updated API endpoints to expose model status information
- Status mapping from Kubernetes conditions to user-friendly status values

<img width="1397" height="355" alt="models" src="https://github.com/user-attachments/assets/0f78bc52-a889-436a-b935-40e1d7479ce3" />

## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via SDKs or endpoints
- [x] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person> for <code, bug, docs, etc>"
- [x] Linked issues #eg101

havasik committed:
feat: simplify model reconcile function

havasik committed:
feat: add conditions to models with refresh interval

havasik committed:
feat(api): handle model conditions instead of phase

havasik committed:
feat(dashboard): handle updated model statuses

havasik committed:
Merge branch 'main' into feat/model-status

havasik committed:
feat: update docs with model changes

havasik committed:
Merge branch 'main' into feat/model-status

dwmkerr commented:
wow this looks amazing, checking now

dwmkerr committed:
Merge branch 'main' into feat/model-status

dwmkerr commented:
Could we do `Ready`, `Pending`, `Error` to be consistent with what we're adding to `model`? 

CC https://github.com/mckinsey/agents-at-scale-ark/pull/104

@adriander FYI this is how it'd look with conditions which are even nicer than the phase which we could move to - we did it for MCP/A2A

@havasik this looks amazing, do you think you could look at some chainsaw tests?

dwmkerr commented on ark/internal/controller/model_controller.go:
```diff
@@ -4,25 +4,30 @@ package controller
 
+const (
+	// Condition types
+	ModelReady       = "Ready"
```
This is where I'd use `Pending` `Ready` `Error` (error might be wrong it may just be pending WDYT though? curious for your thoughts @havasik )

dwmkerr replied:
deployments just stay in pending i think and the events make it clear

havasik replied:
You probably experienced models stuck in "pending" because this fix was not merged in here yet https://github.com/mckinsey/agents-at-scale-ark/pull/122 to refresh sdk on `devspace` (I noticed the same).

havasik replied:
Based on a bit of research on k8s history and maintainer issue discussions, conditions are meant to be orthogonal, easily extensible concepts that can be toggled back and forth individually to signal specific aspects of the resource. The application logic (ark-api in this case) can build on top of this in a more "state machine"-like way, keeping future extensibility of the base spec in mind.

Some references:
- https://github.com/kubernetes/kubernetes/issues/7856
- https://github.com/kubernetes/community/pull/4521

Consequently, by above standards "pending", "ready" and "error" might be not the best choices for condition names, we can just use a single condition since what we're effectively doing here is validating the availability of a specific model. Here's a followup fix to use a simple ModelAvailable condition to determine a "status" label for a given model, I think this is a bit more simplified and better follows the k8s patterns.

https://github.com/mckinsey/agents-at-scale-ark/pull/72/commits/fad1db2ea01af75a34109febec843233bbd2d660

havasik replied:
@dwmkerr I iterated on this PR a bit, based on your helpful comments! 

- updated the Model spec, based on what I found as Kubernetes best practices (detailed above) https://github.com/mckinsey/agents-at-scale-ark/pull/72/commits/fad1db2ea01af75a34109febec843233bbd2d660
- updated the chainsaw tests for the new schema https://github.com/mckinsey/agents-at-scale-ark/pull/72/commits/90582d4332ef7404a3d800e4eb4aeb35b39b5ea1

Let me know what you think about the above guidelines about "conditions", or anything else you think I should add

dwmkerr replied:
perfect!

dwmkerr commented on ark/internal/controller/model_controller.go:
```diff
@@ -32,58 +37,67 @@ type ModelReconciler struct {
+	})
+
+	if err := r.validateModel(ctx, model); err != nil {
+		log.Error(err, "model validation failed", "model", model.Name)
```
might be good to have a few of these: https://mckinsey.github.io/agents-at-scale-ark/developer-guide/logging-and-events/

havasik replied:
I added both logs and events for all of these steps
- Model probing start
- Model probing completion
- Model probing error

dwmkerr reviewed (changes requested):
Frigging amazing - few comments

Later on we could separate out the `v1/completions` check and make it more configurable as embeddings models for example don't have this so at some point we might need something like a model type but I'll spec that first'

This is awesome :)

havasik committed:
Merge branch 'main' into feat/model-status

havasik committed:
fix: update model cards view with new statuses

havasik committed:
fix: noisy model reconciler re-triggers and event failures

dwmkerr committed:
Merge branch 'main' into feat/model-status

havasik committed:
fix: remove unused variable

havasik committed:
Merge branch 'main' into feat/model-status

havasik committed:
refactor: model conditions align to kubernetes design concepts

- a single condition is used to evaluate availability of the referenced model
- api logic reflects this change and exports "pending", "ready" and "error" status labels to dashboard
- tests updated

havasik committed:
fix: update model crd in helm dist

havasik committed:
fix: update chainsaw tests to reflect new Model status schema

havasik committed:
Merge branch 'main' into feat/model-status

havasik committed:
Merge branch 'main' into feat/model-status

dwmkerr commented:
This looks great i'm testing locally @havasik

dwmkerr commented:
This is awesome BTW I'm going to merge and raise a PR to see if we can do similar for the agents

dwmkerr commented:
@havasik love this change so much i did the follow up for agents here #157 and made the UI more consistent with the underlying k8s state, this has *massively* helped it also means i can apply charts like this https://github.com/dwmkerr/agents-at-scale-ark and the eventualy consistently fixes all the race conditions and the agents/models just reconcile properly
